File: go/arraylist/longest_palindrome.go
package arraylist

func LongestPalindrome(s string) string {
	res := ""
	for i := 0; i < len(s); i++ {
		// 以s[i]为中心的最长回文子串
		s1 := palindrome(s, i, i)
		// 以s[i]和s[i+1]为中心的最长回文子串
		s2 := palindrome(s, i, i+1)

		// 取最长回文串
		if len(s1) > len(res) {
			res = s1
		}
		if len(s2) > len(res) {
			res = s2
		}
	}

	return res
}

func palindrome(s string, l, r int) string {
	for l >= 0 && r < len(s) && s[l] == s[r] {
		// 当前是回文，向两边展开，继续判断
		l--
		r++
	}
	// 返回 l, r之间的最长回文
	return s[l+1 : r]
}

File: rust/src/arraylist/longest_palindrome.rs
pub fn longest_palindrome(s: String) -> String {
    let mut res = "";
    for i in 1..s.len() {
        let s1 = palindrome(&s, i as isize , i as isize);
        let s2 = palindrome(&s, i as isize, (i+1) as isize);

        if s1.len() > res.len() {
            res = s1;
        }
        if s2.len() > res.len() {
            res = s2;
        }
    }

    res.to_string()
}

fn palindrome(s: &str, l: isize, r: isize) -> &str {
    let mut l = l;
    let mut r = r;
    let bytes = s.as_bytes();
    while l >= 0 && r < s.len() as isize && bytes[l as usize] == bytes[r as usize] {
        l -= 1;
        r += 1;
    }

    &s[(l+1) as usize..r as usize]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_longest_palindrome() {
        let s = "babad";
        let res = longest_palindrome(s.to_string());
        assert_eq!("bab".to_string(), res);

        let s = "cbbd";
        let res = longest_palindrome(s.to_string());
        assert_eq!("bb".to_string(), res);
    }
}
